pkg/controllermanager/controller/managedseedset/managedseedset_actuator.go

```diff
@@ -0,0 +1,666 @@
+				scheduledShoots = append(scheduledShoots, &shoots.Items[i])
+			}
+		}
+
+		// Add replica
+		replicas = append(replicas, &replica{
+			shoot:           &shoots.Items[i],
+			managedSeed:     managedSeedsByName[shoot.Name],
+			seed:            seedsByName[shoot.Name],
+			scheduledShoots: scheduledShoots,
+		})
+	}
+
+	return replicas, nil
+}
+
+type replicaStatus int
+
+const (
+	statusUnknown               replicaStatus = iota // 0
+	statusShootReconcileFailed                       // 1
+	statusShootDeleteFailed                          // 2
+	statusShootReconciling                           // 3
+	statusShootDeleting                              // 4
+	statusShootReconciled                            // 5
+	statusManagedSeedPreparing                       // 6
+	statusManagedSeedDeleting                        // 7
+	statusManagedSeedRegistered                      // 8
+)
+
+type replica struct {
+	shoot           *gardencorev1beta1.Shoot
+	managedSeed     *seedmanagementv1alpha1.ManagedSeed
+	seed            *gardencorev1beta1.Seed
+	scheduledShoots []*gardencorev1beta1.Shoot
+}
+
+func (r *replica) getShoot() *gardencorev1beta1.Shoot {
+	return r.shoot
+}
+
+func (r *replica) getManagedSeed() *seedmanagementv1alpha1.ManagedSeed {
+	return r.managedSeed
+}
+
+func (r *replica) getObjectName() string {
+	return kutil.ObjectName(r.shoot)
+}
+
+// getOrdinal gets the replica's ordinal. If the replica has no ordinal, -1 is returned.
+func (r *replica) getOrdinal() int {
+	_, ordinal := getParentNameAndOrdinal(r.shoot.Name)
+	return ordinal
+}
+
+// getStatus gets the replica's status.
+func (r *replica) getStatus() replicaStatus {
+	switch {
+	case r.shoot != nil && r.managedSeed == nil:
+		switch {
+		case shootReconcileSucceeded(r.shoot):
+			return statusShootReconciled
+		case shootReconcileFailed(r.shoot):
+			return statusShootReconcileFailed
+		case shootDeleteFailed(r.shoot):
+			return statusShootDeleteFailed
+		case r.shoot.DeletionTimestamp == nil:
+			return statusShootReconciling
+		default:
+			return statusShootDeleting
+		}
+	case r.shoot != nil && r.managedSeed != nil:
+		switch {
+		case managedSeedRegistered(r.managedSeed):
+			return statusManagedSeedRegistered
+		case r.managedSeed.DeletionTimestamp == nil:
+			return statusManagedSeedPreparing
+		default:
+			return statusManagedSeedDeleting
+		}
+	default:
+		return statusUnknown
+	}
+}
+
+func (r *replica) isSeedReady() bool {
+	return r.seed != nil && seedReady(r.seed)
+}
+
+func (r *replica) getShootHealthStatus() operationshoot.Status {
+	return shootHealthStatus(r.shoot)
+}
+
+func (r *replica) isReady() bool {
+	return r.getStatus() == statusManagedSeedRegistered && r.isSeedReady() && r.getShootHealthStatus() == operationshoot.StatusHealthy
+}
+
+func (r *replica) isDeletable() bool {
+	return len(r.scheduledShoots) == 0 &&
+		!kutil.HasMetaDataAnnotation(r.shoot, seedmanagementv1alpha1constants.AnnotationProtectFromDeletion, "true")
```

Comment:
Why did you chose to put the protection annotation on the shoot?
Without much thought, it seems to make more sense to have it on the `ManagedSeed`, as the `ManagedSeedSet` manages individual `ManagedSeed`s.

Reply:
Because the shoot is the only replica object guaranteed to exist at all times. If I want to protect a half-created replica from deletion, the only object that is guaranteed to be there is the shoot.

Reply:
We agreed to support this annotation also on managed seeds (in addition to shoots), I will add this.